### ISSUE TITLE
SSD7 Gives Negatives Coordinates As Outputs

### ISSUE THREAD

**Issue by chensong1995** [pierluigiferrari/ssd_keras#134]
Hi there,

I am trying to train an ssd7 to detect cat drawings in 300*300 grayscale images, and I am confused when I get negative bounding box coordinates as outputs.

An example of the image looks like this:
![Training example](http://songc.me/doc/0.bmp "Training example")

The bounding box for that particular example is given in the csv format as:
```
C:\Users\csa102\Documents\USRA\GAN-logo\cat-text\mix-big\0.bmp, 1, 62, 155, 144, 237
```

I am basically following the [SSD7 Training Tutorial](https://github.com/pierluigiferrari/ssd_keras/blob/master/ssd7_training.ipynb). However, it appears that the existing code assumes that input has always 3 channels. In order to make it work with grayscale images, the changes I have to make are:

1. Disabling HDF5 datasets
2. Disabling `data_augmentation_chain`
3. In [data_generator/object_detection_2d_data_generator.py](https://github.com/pierluigiferrari/ssd_keras/blob/master/data_generator/object_detection_2d_data_generator.py#L1017), at line 1017 I have to reshape `image` from (300, 300) to (300, 300, 1)

This is the code changes I made in `object_detection_2d_data_generator.py`.
```python
with Image.open(filename) as image:
    batch_X.append(np.array(image, dtype=np.uint8))
    image_np = np.array( image, dtype=np.uint8 )
    if image_np.ndim == 2:
        image_np = np.reshape( image_np, image_np.shape + ( 1, ) )
    batch_X.append( image_np )
```

After training, I tried using the model to do predictions on the validation set:

```
Image: C:\Users\csa102\Documents\USRA\GAN-logo\cat-text\mix-big\15834.bmp

Ground truth boxes:
[[  1  72 135 121 184]]

Predicted boxes:
   class   conf xmin   ymin   xmax   ymax
[[   1.      1.    130.84  222.32  159.92  257.92]
 [   1.      1.     67.74  219.37  106.71  262.79]
 [   1.      1.    142.02  144.48  179.61  183.94]
 [   1.      1.     69.48  151.31  110.75  193.7 ]
 [   1.      1.    119.72   43.96  163.17   92.18]
 [   1.      1.    204.87  153.86  236.78  186.73]
 [   1.      1.     27.83  133.11   74.56  181.59]
 [   1.      1.     51.33   35.75   95.82   90.55]
 [   1.      1.    181.01   83.77  222.62  125.27]
 [   1.      1.     32.28  194.41   70.3   237.21]
 [   1.      0.99   88.43  248.76  119.6   282.54]
 [   1.      0.99  198.96  208.13  224.61  238.4 ]
 [   1.      0.97  136.4   238.78  163.23  271.52]
 [   1.      0.67   32.14  102.24  262.61  349.49]
 [   1.      0.45  -40.56  -44.94  341.84  328.19]
 [   1.      0.32 -496.74  112.46  790.41  185.74]
 [   1.      0.26   76.38   86.44  118.06  130.93]
 [   1.      0.23   45.94   64.08   93.11  115.94]
 [   1.      0.15  124.54  111.59  168.88  150.62]]
```

Some of the coordinates are clearly outside the (300, 300) region. I suspect that may be I have given wrong arguments to the encoder and/or decoder.

Here is a summary of my environment:
* __Operating System:__ Windows 10
* __Which commit of this repository I am on:__ [the lastest as of June 10, 2018](https://github.com/pierluigiferrari/ssd_keras/commit/3ac9adaf3889f1020d74b0eeefea281d5e82f353)
* __Keras version__: 2.1.6
* __TensorFlow version__: 1.8.0

The complete code I used for training is pasted below for reference. Any help or suggestion will be greatly appreciated.

```python
from keras.optimizers import Adam
from keras.callbacks import ModelCheckpoint, EarlyStopping, ReduceLROnPlateau, TerminateOnNaN, CSVLogger
from keras import backend as K
from keras.models import load_model
from math import ceil
import numpy as np
from matplotlib import pyplot as plt

from models.keras_ssd7 import build_model
from keras_loss_function.keras_ssd_loss import SSDLoss
from keras_layers.keras_layer_AnchorBoxes import AnchorBoxes
from keras_layers.keras_layer_DecodeDetections import DecodeDetections
from keras_layers.keras_layer_DecodeDetectionsFast import DecodeDetectionsFast

from ssd_encoder_decoder.ssd_input_encoder import SSDInputEncoder
from ssd_encoder_decoder.ssd_output_decoder import decode_detections, decode_detections_fast

from data_generator.object_detection_2d_data_generator import DataGenerator
from data_generator.object_detection_2d_misc_utils import apply_inverse_transforms
from data_generator.data_augmentation_chain_variable_input_size import DataAugmentationVariableInputSize
from data_generator.data_augmentation_chain_constant_input_size import DataAugmentationConstantInputSize
from data_generator.data_augmentation_chain_original_ssd import SSDDataAugmentation

import pdb

img_height = 300
img_width = 300
img_channels = 1
n_classes = 1
normalize_coords = True

# Build model
K.clear_session() # Clear previous models from memory.
model = build_model( image_size=( img_height, img_width, img_channels ),
                     n_classes=n_classes,
                     mode='training',
                     l2_regularization=0.0005,
                     clip_boxes=True,
                     normalize_coords=normalize_coords )

adam = Adam( lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0 )
ssd_loss = SSDLoss( neg_pos_ratio=3, alpha=1.0 )
model.compile( optimizer=adam, loss=ssd_loss.compute_loss )

# Configure datasets
train_dataset = \
        DataGenerator( load_images_into_memory=False, hdf5_dataset_path=None )
val_dataset = DataGenerator( load_images_into_memory=False, hdf5_dataset_path=None )

images_dir = r'C:\Users\csa102\Documents\USRA\GAN-logo\cat-text\mix-big'
train_labels_filename = r'C:\Users\csa102\Documents\USRA\GAN-logo\cat-text\labels_big_train.csv'
val_labels_filename   = r'C:\Users\csa102\Documents\USRA\GAN-logo\cat-text\labels_big_val.csv'
input_format = [ 'image_name', 'class_id', 'ymin', 'ymax', 'xmin', 'xmax' ]
train_dataset.parse_csv( images_dir=images_dir,
                         labels_filename=train_labels_filename,
                         input_format=input_format,
                         include_classes='all' )
val_dataset.parse_csv( images_dir=images_dir,
                       labels_filename=val_labels_filename,
                       input_format=input_format,
                       include_classes='all' )

train_dataset_size = train_dataset.get_dataset_size()
val_dataset_size   = val_dataset.get_dataset_size()
print( "Number of images in the training dataset:\t{:>6}".\
        format( train_dataset_size ) )
print( "Number of images in the validation dataset:\t{:>6}".\
        format( val_dataset_size ) )

batch_size = 32

# Data pre-processers
predictor_sizes = [ model.get_layer( 'classes4' ).output_shape[ 1 : 3 ],
                    model.get_layer( 'classes5' ).output_shape[ 1 : 3 ],
                    model.get_layer( 'classes6' ).output_shape[ 1 : 3 ],
                    model.get_layer( 'classes7' ).output_shape[ 1 : 3 ] ]
ssd_input_encoder = SSDInputEncoder( img_height=img_height,
                                     img_width=img_width,
                                     n_classes=n_classes,
                                     predictor_sizes=predictor_sizes,
                                     clip_boxes=True,
                                     matching_type='multi',
                                     pos_iou_threshold=0.5,
                                     neg_iou_limit=0.3,
                                     normalize_coords=normalize_coords )

train_generator = train_dataset.generate( batch_size=batch_size,
                                          shuffle=True,
                                          transformations=[],
                                          label_encoder=ssd_input_encoder,
                                          returns={ 'processed_images',
                                                    'encoded_labels' },
                                          keep_images_without_gt=False )
val_generator = val_dataset.generate( batch_size=batch_size,
                                      shuffle=False,
                                      transformations=[],
                                      label_encoder=ssd_input_encoder,
                                      returns={ 'processed_images',
                                                'encoded_labels' },
                                      keep_images_without_gt=False )

# Training callbacks
model_checkpoint = ModelCheckpoint( filepath='ssd7_epoch-{epoch:02d}_loss-{loss:.4f}_val_loss-{val_loss:.4f}.h5',
                                    monitor='val_loss',
                                    verbose=1,
                                    save_best_only=True,
                                    save_weights_only=False,
                                    mode='auto',
                                    period=1 )
csv_logger = CSVLogger( filename='ssd7_training_log.csv',
                        separator=',',
                        append=True )
early_stopping = EarlyStopping( monitor='val_loss',
                                min_delta=0.0,
                                patience=10,
                                verbose=1 )
reduce_learning_rate = ReduceLROnPlateau( monitor='val_loss',
                                          factor=0.2,
                                          patience=8,
                                          verbose=1,
                                          epsilon=0.001,
                                          cooldown=0,
                                          min_lr=0.00001 )
callbacks = [ model_checkpoint,
              csv_logger,
              early_stopping,
              reduce_learning_rate ]


# Training
initial_epoch   = 0
final_epoch     = 100
steps_per_epoch = 1000

history = model.fit_generator( generator=train_generator,
                               steps_per_epoch=steps_per_epoch,
                               epochs=final_epoch,
                               callbacks=callbacks,
                               validation_data=val_generator,
                               validation_steps=ceil( val_dataset_size / batch_size ),
                               initial_epoch=initial_epoch )

# Predication
predict_generator = val_dataset.generate( batch_size=1,
                                          shuffle=True,
                                          transformations=[],
                                          label_encoder=None,
                                          returns={ 'processed_images',
                                                    'processed_labels',
                                                    'filenames' },
                                          keep_images_without_gt=False )

batch_images, batch_labels, batch_filenames = next( predict_generator )
i = 0 # Which batch item to look at
print( "Image:", batch_filenames[ i ] )
print()
print( "Ground truth boxes:\n" )
print( batch_labels[ i ] )

y_pred = model.predict( batch_images )
pdb.set_trace()
y_pred_decoded = decode_detections( y_pred,
                                    confidence_thresh=0.1,
                                    iou_threshold=0.45,
                                    top_k=200,
                                    normalize_coords=normalize_coords,
                                    img_height=img_height,
                                    img_width=img_width )

np.set_printoptions( precision=2, suppress=True, linewidth=90 )
print( "Predicted boxes:\n" )
print( '   class   conf xmin   ymin   xmax   ymax' )
print( y_pred_decoded[ i ] )
```

**Comment by pierluigiferrari** [comment]
A few things:
1. I generally don't debug user code (including user modifications to my code), no time for that.
2. Why did you go through so much trouble to avoid the conversion to 3 color channels? What's the downside of the greyscale images being converted to 3 channels? You could have just left everything as is and it would have worked fine with your greyscale images.
3. Regarding the negative coordinates:
   1. Some of these predicted boxes definitely don't make any sense, but note that it is generally perfectly possible for the model to predict negative numbers. The model doesn't care about image boundaries. If the borders of an object are predicted to lie beyond the image borders, then that's what the model will give you.
   2. The CSV input coordinate format you stated above doesn't say anything. It's just a bunch of numbers. It doesn't say which number represents which coordinate.
   3. With the previous point being said, make sure you set the input coordinate format correctly, otherwise the model can't learn anything sensical.

**Comment by chensong1995** [comment]
Thanks very much for you comment.

*The reason why I am avoiding the conversion to 3 color channels is that I want to simplify the internal model structure. 
* The CSV format is given by `input_format = [ 'image_name', 'class_id', 'ymin', 'ymax', 'xmin', 'xmax' ]` in the training code. I am sure it is correct.
* I thought the decoding method would always make sure the output coordinates are within the image boundaries. If, as you said, that this is not true, then perhaps SSD is not suitable for my particular task? The object I want to detect does not have any texture or color, which is very different from PASCAL VOC.

**Comment by pierluigiferrari** [comment]
> *The reason why I am avoiding the conversion to 3 color channels is that I want to simplify the internal model structure.

That makes sense, but you could have started with the 3-channel version and once that would have worked, you could have made the code changes to keep everything 1-channel. This would have given you more control over where the problem might lie. As a general piece of advice, I would recommend to always make as few simultaneous changes as possible to a working system and instead make changes step by step to better be able to see where exactly something breaks.

> I thought the decoding method would always make sure the output coordinates are within the image boundaries.

It doesn't, and as explained above, neither is this necessary (and neither do other object detectors like Faster R-CNN, for example). Does the documentation claim anywhere that it does? If yes, let me know, then I'll have to correct that.

> If, as you said, that this is not true, then perhaps SSD is not suitable for my particular task? The object I want to detect does not have any texture or color, which is very different from PASCAL VOC.

I don't see what clipping bounding box coordinates at the image boundaries has to do with a model's suitability for a given detection task.

Anyways, we're digressing. A simple first step to debug your problem is to have the data generator generate a batch and then display an image in the batch and print the generated ground truth boxes onto the image (the training notebooks do that). Are the ground truth boxes where they are supposed to be? This is to make sure that the ground truth data is being handled correctly within the data generator.

**Comment by chensong1995** [comment]
Thanks again for your comment.

* I am very sure that the ground truth boxes are provided in the correct format. I tried drawing ground truth bounding boxes on training examples, and yes, the boxes are where they are supposed to be.
* I also tried the 3-channel version. The result I got is identical to the 1-channel model: many boxes are detected with high confidences, but the coordinates do not make any sense at all.
* No, the documentation does not claim that the output coordinates are within the image boundaries. It is just my intuitive feeling which proves to be wrong.
* I am suspicious of the suitability of SSD for my specific task not because of the clipping issue, but because the data set I use is very different from the examples in your documentation.

**Comment by chensong1995** [comment]
The results are draw below. The green box is the ground truth which is in the correct position, and the blue box is the predicted box.

![Results](https://user-images.githubusercontent.com/21059441/41372344-e1f3c226-6f01-11e8-9eac-55f3f65aedc6.png)

**Comment by pierluigiferrari** [comment]
So this is the result after 100,000 training steps?

> I am suspicious of the suitability of SSD for my specific task not because of the clipping issue, but because the data set I use is very different from the examples in your documentation.

I mean, it's an object detector. It learns to recognize patterns in images, just like any other object detector. The shapes and sizes of the objects seem pretty standard, so I don't see why it wouldn't be able to learn this task. I can't explain the results you posted above though.

By the way, not that it's going to fix your problem, but:

- Why didn't you pass any scaling factors to `build_model()` and to the `SSDInputEncoder`?
- Why did you set `clip_boxes = True`? It should be `False`.

**Comment by chensong1995** [comment]
Thanks so much! Passing scaling factors to `build_model()` and `SSDInputEncoder`, and also unset `clip_boxes` fixed the problem.

**Comment by pierluigiferrari** [comment]
Oh damn, then I must have a bug in there after all. It's supposed to work even if you don't pass any explicit scaling factors.

Anyways, thanks for reporting back and I'm glad it works now. I'll fix the bug where it doesn't work if you don't supply explicit scaling factors.
